flatsurf/geometry/hyperbolic.py

```diff
@@ -0,0 +1,729 @@
+            sage: H = HyperbolicPlane(QQ)
+            sage: H(H.an_element()) in H
+            True
+
+        """
+        if x.parent() is self:
+            return x
+
+        # TODO: Accept elements that convert into the base ring.
+
+        # TODO: Change ring otherwise.
+
+        raise NotImplementedError
+
+    def base_ring(self):
+        r"""
+        Return the base ring over which objects in the plane are defined.
+
+        EXAMPLES::
+
+            sage: from flatsurf.geometry.hyperbolic import HyperbolicPlane
+
+            sage: HyperbolicPlane().base_ring()
+            Rational Field
+
+        """
+        return self._base_ring
+
+    def infinity(self):
+        r"""
+        Return the point at infinity in the Poincaré half plane model.
+
+        EXAMPLES::
+
+            sage: from flatsurf.geometry.hyperbolic import HyperbolicPlane
+
+            sage: HyperbolicPlane().infinity()
+            ∞
+
+        """
+        return self.projective(1, 0)
+
+    def real(self, r):
+        r"""
+        Return the ideal point ``r`` on the real axis in the Poincaré half
+        plane model.
+
+        EXAMPLES::
+
+            sage: from flatsurf.geometry.hyperbolic import HyperbolicPlane
+
+            sage: HyperbolicPlane().real(-2)
+            -2
+
+        """
+        return self.projective(r, self.base_ring().one())
+
+    def projective(self, p, q):
+        r"""
+        Return the ideal point with projective coordinates ``[p: q]`` in the
+        Poincaré half plane model.
+
+        EXMAPLES::
+
+            sage: from flatsurf.geometry.hyperbolic import HyperbolicPlane
+
+            sage: H = HyperbolicPlane()
+            sage: H.projective(0, 1)
+            0
+
+            sage: H.projective(-1, 0)
+            ∞
+
+            sage: H.projective(0, 0)
+            Traceback (most recent call last):
+            ...
+            ValueError: one of p and q must not be zero
+
+        """
+        if p == 0 and q == 0:
+            raise ValueError("one of p and q must not be zero")
+
+        if q == 0:
+            return self.point(0, 1, model="klein")
+
+        return self.point(p/q, 0, model="half_plane")
+
+    def point(self, x, y, model):
+        r"""
+        Return the point with coordinates (x, y) in the given model.
+
+        When ``model`` is ``"half_plane"``, return the point `x + iy` in the upper half plane.
+
+        When ``model`` is ``"klein"``, return the point (x, y) in the Klein disc.
+
+        EXAMPLES::
+
+            sage: from flatsurf.geometry.hyperbolic import HyperbolicPlane
+            sage: H = HyperbolicPlane()
+
+            sage: H.point(0, 1, model="half_plane")
+            I
+
+            sage: H.point(1, 2, model="half_plane")
+            1 + 2*I
+
+            sage: H.point(0, 1, model="klein")
+            ∞
+
+        """
+        if model == "klein":
+            return self.__make_element_class__(HyperbolicPoint)(self, x, y)
+        if model == "half_plane":
```

Comment:
From what I mentioned above I don't believe this is correct.

Reply:
A good check is to pick points on a geodesic in the upper half-plane and check whether they are aligned in the Klein model. For fixed `a, b, c, d` with `ad - bc ≠ 0`, the image of the vertical geodesic by `[a, b; c, d]` is parametrized by the points `(a t z + b) / (ct z + d)`.